crates/ty_python_semantic/resources/mdtest/generics/legacy/classes.md

```diff
@@ -379,6 +379,59 @@ C[None](b"bytes")  # error: [no-matching-overload]
 C[None](12)
 ```
 
+### Inferring callable return type from generic class constructor
+
+```py
+from typing import Callable, Generic, overload
+from typing_extensions import TypeVar, reveal_type
+
+T = TypeVar("T")
+
+class A(Generic[T]):
+    def __init__(self, x: Callable[..., T]):
+        self.x = x
+
+def f() -> int:
+    return 1
+
+reveal_type(A(f))  # revealed: A[int]
+
+@overload
+def g(x: int) -> int: ...
+@overload
+def g(x: str) -> str: ...
+def g(x: str | int) -> str | int:
+    return x
+
+reveal_type(A(g))  # revealed: A[int | str]
+```
+
+Using a constrained typevar for the callable return type:
+
+```py
+from typing import Callable, Generic, overload
+from typing_extensions import TypeVar, reveal_type
+
+T = TypeVar("T", int, str)
+
+class B(Generic[T]):
+    def __init__(self, x: Callable[..., T | None]):
+        self.x = x
+
+def h(x: int) -> int | None:
+    return x
+
+@overload
+def h(x: str) -> str | None: ...
+@overload
+def h(x: int) -> int | None: ...
+def h(x: str | int) -> str | int | None:
+    return x
+
+# TODO: should be B[int | str]
```

Comment:
Hmm on second look I don't think `B[int | str]` is right here. The typevar is constrained to `int` and `str`, so we must solve it to one of those types. This makes me think that the overloaded function `h` can't be accepted here, and we should actually be erroring on `B(h)`, because we can't solve the typevar to either `int` or `str`.

It would probably be good to also have a test with constrained typevar that doesn't use an overloaded callable, just a normal callable returning e.g. `int`, and show we can solve that to `int`.

Reply:
For what it's worth, pyright reveals `B[str]` for this example, presumably since the first overload matches?

Reply:
Hmm, I guess that makes sense. It probably isn't right to error, since the given callable does satisfy the annotation; it's just ambiguous whether we should infer `str` or `int`. But I guess in the face of that ambiguity, choosing the first overload is reasonable.